tests/unit/store/materialize.ts

```diff
@@ -0,0 +1,263 @@
+			.transform((item) => ({
+				_value: item.value,
+				_nestedProperty: {
+					_value: item.nestedProperty.value
+				},
+				_id: item.id
+			}), '_id')
+			.filter(function(item) {
+				return item._value > 1;
+			})
+			.range(0, 100)
+			.track();
+
+		let ignoreFirst = true;
+		targetStore.observe().subscribe(dfd.rejectOnError((update: StoreDelta<any>) => {
+			if (ignoreFirst) {
+				ignoreFirst = false;
+				return;
+			}
+			dfd.reject(Error('Shouldn\'t have received another update since apply function was used'));
+		}));
+
+		let initialUpdateFromSource = false;
+		materialize({
+			source: trackedCollection,
+			target: targetStore,
+			apply: dfd.rejectOnError((target: Store<any, any, any>, { afterAll, deletes }: StoreDelta<any>, source: any) => {
+				if (!initialUpdateFromSource) {
+					initialUpdateFromSource = true;
+					assert.deepEqual(afterAll, [
+						{
+							_id: '2',
+							_value: 2,
+							_nestedProperty: {
+								_value: 2
+							}
+						},
+						{
+							_id: '3',
+							_value: 3,
+							_nestedProperty: {
+								_value: 1
+							}
+						}
+					], 'Didn\'t add initial data to target store');
+				}
+				else {
+					assert.deepEqual(deletes, [ '2' ]);
+					dfd.resolve();
+				}
+			})
+		});
+
+		trackableQueryStore.delete('2');
+	},
+
+	'Shouldn\'t make any updates if initial update is empty'(this: any) {
+		const dfd = this.async(1000);
+		const targetStore = createQueryStore<ItemType>();
+		const trackableQueryStore = createQueryStore<ItemType>();
+		const trackedCollection = trackableQueryStore.filter(() => true).track();
+
+		targetStore.add = dfd.reject.bind(dfd, Error('Shouldn\'t have called add on targetStore'));
+
+		materialize({
+			source: trackedCollection,
+			target: targetStore
+		});
+
+		setTimeout(dfd.resolve, 200);
+	},
+
+	'Should add new items to targetStore'(this: any) {
+		const dfd = this.async(1000);
+		const targetStore = createQueryStore<ItemType>();
+		const trackableQueryStore = createQueryStore<ItemType>();
+		const trackedCollection = trackableQueryStore.filter(() => true).track();
+
+		const spy = sinon.spy(targetStore, 'add');
+
+		materialize({
+			source: trackedCollection,
+			target: targetStore
+		});
+
+		trackableQueryStore.add(createData());
+
+		setTimeout(dfd.callback(() => {
+			assert.isTrue(spy.calledOnce, 'Should have called add on target store once');
+			assert.deepEqual(spy.args[0][0], createData());
+		}), 100);
+	},
+
+	'Should update items in target store'(this: any) {
+		const dfd = this.async(1000);
+		const targetStore = createQueryStore<ItemType>();
+		const trackableQueryStore = createQueryStore<ItemType>();
+		const trackedCollection = trackableQueryStore.filter(() => true).track();
+
+		const spy = sinon.spy(targetStore, 'put');
+
+		materialize({
+			source: trackedCollection,
+			target: targetStore
+		});
+
+		trackableQueryStore.add(createData());
+		trackableQueryStore.put(createUpdates()[0]);
+
+		setTimeout(dfd.callback(() => {
```

Comment:
we can use `dojo-core/timing#delay` for these scenarios and return the promise instead of using the `dfd.callback`